### PR TITLE
ACCUMULO-4419: Change how compression delegation works

### PR DESCRIPTION
**Pull request by phrocker**
This commit includes A compression pool not based on CodecPool,
a default factory that does no pooling. Changes are also
in place that will allow us to better control some aspects of
Compression within the tserver

These are early changes, so I'm soliciting feedback before I continue. My main concerns are:

Do we really need the ability to trim the codec pool and the other features that commons-pool provides.
Does this really need to be configurable? I believe so as we are using this.
What additional tests should I add? I'm working a few more but if there ideas I would gladly ignore them ;)
Should I make anything else configurable for this first pass? 
When scrolling through the changes there are some comments I'd like to see so I'm going to update comments. If you have any input I'm all ears.

### CODE REVIEW

File: core/src/main/java/org/apache/accumulo/core/conf/Property.java
Comment (phrocker) [inline core/src/main/java/org/apache/accumulo/core/conf/Property.java]: More ideas for configuration please.

File: core/src/main/java/org/apache/accumulo/core/file/rfile/bcfile/codec/CompressionUpdater.java
Comment (joshelser) [inline core/src/main/java/org/apache/accumulo/core/file/rfile/bcfile/codec/CompressionUpdater.java]: Remove.

File: core/src/main/java/org/apache/accumulo/core/file/rfile/bcfile/Compression.java
Comment (phrocker) [inline core/src/main/java/org/apache/accumulo/core/file/rfile/bcfile/Compression.java]: I'm going to move the conditional to separate lines to improve readability

File: core/src/main/java/org/apache/accumulo/core/file/rfile/bcfile/codec/CompressionUpdater.java
Comment (phrocker) [inline core/src/main/java/org/apache/accumulo/core/file/rfile/bcfile/codec/CompressionUpdater.java]: oh noes! thanks

**Comment by phrocker** [comment]
I didn't build the java doc. there are errors. I'm going to correct and re-submit. In the mean time please send along any comments.

File: core/src/main/java/org/apache/accumulo/core/file/rfile/bcfile/codec/CompressorFactory.java
Comment (joshelser) [inline core/src/main/java/org/apache/accumulo/core/file/rfile/bcfile/codec/CompressorFactory.java]: Should this be an interface if it's a potential extension point?

File: core/src/main/java/org/apache/accumulo/core/file/rfile/bcfile/codec/CompressorObjectFactory.java
Comment (joshelser) [inline core/src/main/java/org/apache/accumulo/core/file/rfile/bcfile/codec/CompressorObjectFactory.java]: Naming seems a little weird. "CompressorObjectFactory" and "CompressorFactory" seem to overlap in functionality based on the class name.

File: core/src/main/java/org/apache/accumulo/core/file/rfile/bcfile/codec/CompressorPool.java
Comment (joshelser) [inline core/src/main/java/org/apache/accumulo/core/file/rfile/bcfile/codec/CompressorPool.java]: Javadoc, please.

File: core/src/main/java/org/apache/accumulo/core/file/rfile/bcfile/codec/CompressorPool.java
Comment (joshelser) [inline core/src/main/java/org/apache/accumulo/core/file/rfile/bcfile/codec/CompressorPool.java]: Commons-Pool doesn't give us a named exception to differentiate between the Pool failing to construct a new instance and some pool operation failing? (your `catch Exception` clause)

File: core/src/main/java/org/apache/accumulo/core/file/rfile/bcfile/codec/CompressorPool.java
Comment (joshelser) [inline core/src/main/java/org/apache/accumulo/core/file/rfile/bcfile/codec/CompressorPool.java]: An associated error message would be good.

File: core/src/test/java/org/apache/accumulo/core/file/rfile/bcfile/codec/CompressorFactoryTest.java
Comment (joshelser) [inline core/src/test/java/org/apache/accumulo/core/file/rfile/bcfile/codec/CompressorFactoryTest.java]: nit, spelling

File: core/src/test/java/org/apache/accumulo/core/file/rfile/bcfile/codec/CompressorFactoryTest.java
Comment (joshelser) [inline core/src/test/java/org/apache/accumulo/core/file/rfile/bcfile/codec/CompressorFactoryTest.java]: Could lift the null && true check into a method for clarity since it's copied in a couple of places.

File: pom.xml
Comment (joshelser) [inline pom.xml]: Should we be trying to use org.apache.commons:commons-pool2 instead? I'm not sure if commons-pool is still receiving updates.

File: core/src/main/java/org/apache/accumulo/core/file/rfile/bcfile/Compression.java
Comment (joshelser) [inline core/src/main/java/org/apache/accumulo/core/file/rfile/bcfile/Compression.java]: I think making this a `final AtomicReference<CompressorFactory>` would be good (corresponding comment below).

File: core/src/main/java/org/apache/accumulo/core/file/rfile/bcfile/Compression.java
Comment (joshelser) [inline core/src/main/java/org/apache/accumulo/core/file/rfile/bcfile/Compression.java]: With the `AtomicReference`, we could remove the synchronization and just close the previous value we get from the `getAndSet(V)` method.

**Comment by joshelser** [comment]
On overall comment: let's use slf4j loggers instead of log4j specifically. I see you have log4j classes sprinkled around.

File: pom.xml
Comment (ctubbsii) [inline pom.xml]: Yes, I think we should. For a new dependency, I think we should be using the latest available, to ensure the best upstream support.

Comment (phrocker) [inline pom.xml]: Agreed. Not sure why I chose 1.6  when I have been using Commons pool2

File: core/src/main/java/org/apache/accumulo/core/file/rfile/bcfile/Compression.java
Comment (phrocker) [inline core/src/main/java/org/apache/accumulo/core/file/rfile/bcfile/Compression.java]: Ahh yah.... Good call.

File: core/src/main/java/org/apache/accumulo/core/conf/Property.java
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/conf/Property.java]: Would it make sense to drop this config and automatically derive it as a function of `tserver.compressor.pool.max.idle.time`?

File: core/src/main/java/org/apache/accumulo/core/conf/Property.java
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/conf/Property.java]: Was compressor pooing was unlimited before this change?  If so, is this default too low (I am not saying it is, I have no idea)?  How much time does it take to allocate a decompressor?  I am wondering about a case where I get a burst of scans like 300 and then a few seconds later get another burst, do you have a sense of how that scenario perform before and after this change w/ this default?

File: core/src/test/java/org/apache/accumulo/core/file/rfile/bcfile/codec/CompressorFactoryTest.java
Comment (keith-turner) [inline core/src/test/java/org/apache/accumulo/core/file/rfile/bcfile/codec/CompressorFactoryTest.java]: OMG I found a spelling mistake.   I am usually the one generating spelling errors.

File: core/src/main/java/org/apache/accumulo/core/file/rfile/bcfile/Compression.java
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/file/rfile/bcfile/Compression.java]: I was trying to figure out how the following situation will work out by following the code.  But I got lost and decided to just ask.

 1. Thread 1 gets decompressor D1 from compression factory F1
 1. Thread 2 sets the compression factory to F2
 1. Thread returns its decompressor D1. Does it go to F1 or F2?

File: core/src/test/java/org/apache/accumulo/core/file/rfile/bcfile/codec/CompressorFactoryTest.java
Comment (ctubbsii) [inline core/src/test/java/org/apache/accumulo/core/file/rfile/bcfile/codec/CompressorFactoryTest.java]: Are you sure that's a spelling mistake? Maybe it was intended to read as "Al Gore items", with a lisp. ;)

File: core/src/main/java/org/apache/accumulo/core/file/rfile/bcfile/Compression.java
Comment (phrocker) [inline core/src/main/java/org/apache/accumulo/core/file/rfile/bcfile/Compression.java]: Implementations close compressors if they don't own it. I realize the fragility in making sure the release call closes the compressor/decompressor. The alternative I imagined would return a status from release that shows us whether or not the object was released. in that case Compression would call end manually if the factory does not.

Comment (phrocker) [inline core/src/main/java/org/apache/accumulo/core/file/rfile/bcfile/Compression.java]: I'm gonna add some tests to prove this. Also address the performance questions from earlier.

Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/file/rfile/bcfile/Compression.java]: > Implementations close compressors if they don't own it.

Thanks for the explanation.  Thats a nice simple approach.  

> I realize the fragility in making sure the release call closes the compressor/decompressor. 

I would rather the pool close it when it does not own it as you suggested and document this expectation in the base class release method javadoc.

**Comment by phrocker** [comment]
This one fell off of my radar. will be posting an update today some time.

**Comment by asfgit** [comment]
Can one of the admins verify this patch?

**Comment by milleruntime** [comment]
@phrocker are you planning on finishing this any time soon?  If not, I will close the PR for now.

**Comment by phrocker** [comment]
@milleruntime it was finished, but not merged. If I have an opportunity to make sure it works I'll merge. Thanks.

**Comment by milleruntime** [comment]
OK sounds good.  If you don't get to it over the next week, maybe someone else can test it.

@keith-turner were all your questions addressed?

**Comment by milleruntime** [comment]
Closing this PR since the work is complete.

**Comment by ctubbsii** [comment]
This PR was never merged into the code base and its corresponding JIRA issue remains open.

**Comment by ctubbsii** [comment]
asfgit closed this because the 1.7 branch was deleted. I've changed the base branch to master now. If the work is continued, it should be done against an active branch.

**Comment by Fja1975** [comment]
Yo

**Comment by phrocker** [comment]
Taking a fresh look at this now.

**Comment by phrocker** [comment]
@ctubbsii and @keith-turner  thanks for prior reviews. I've rebased and am testing this on a cluster. I built and ran tests locally but did not run verify. I will do that now and allow travis to do its thing. I'll post an update when it's ready for another set of eyes, which I would totally appreciate. Sorry for the mixup on my part. This was a PR that had a positive impact on production systems -- I just became too peanut buttered.

**Comment by phrocker** [comment]
Hey @keith-turner , if you have a chance can you take a quick look at this again? I made updates and added a test. the FX hasn't changed but given that there have been so many comments I wanted to get a quick run through by you if you have a chance before merging this. thanks!

**Comment by keith-turner** [comment]
Sure thing @phrocker. I am currently looking at it.

File: core/src/main/java/org/apache/accumulo/core/conf/Property.java
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/conf/Property.java:530]: I think these are options for a specific plugin.  If someone configures another plugin class, then these options may not be relevant.  Could possibly structure props like :

```
tserver.compressor.factory.class=<class>
tserver.compressor.factory.opts.<key1>=<value1>
tserver.compressor.factory.opts.<key2>=<value2>
```

And the plugin would be given a map containing `<key1>=<value1>,<key2>=<value2>`

File: core/src/main/java/org/apache/accumulo/core/file/rfile/bcfile/codec/CompressorPool.java
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/file/rfile/bcfile/codec/CompressorPool.java:27]: I see this is in an impl package of commons.pool2. I wonder if that implies it may not be stable.

File: core/src/main/java/org/apache/accumulo/core/file/rfile/bcfile/codec/CompressorPool.java
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/file/rfile/bcfile/codec/CompressorPool.java:56]: I looked at the commons pool javadoc a bit and noticed it had a GenericKeyedObjectPoolConfig object.  Not sure if this is workable, but I was wondering about making the config pass through.  Where a prop could be set in the Accumulo config for any setter on GenericKeyedObjectPoolConfig.  Like GenericKeyedObjectPoolConfig has a setMaxTotal method. The accumulo config could have something like.

```
tserver.compressor.factory.class=...CompressorPool
tserver.compressor.factory.opts.maxTotal=42
```

and here when it sees maxTotal it calls the setMaxTotal method on GenericKeyedObjectPoolConfig.

File: core/src/main/java/org/apache/accumulo/core/file/rfile/bcfile/codec/CompressorPool.java
Comment (keith-turner) [reply core/src/main/java/org/apache/accumulo/core/file/rfile/bcfile/codec/CompressorPool.java:98]: In [this javadoc](https://commons.apache.org/proper/commons-pool/api-2.8.0/org/apache/commons/pool2/impl/GenericKeyedObjectPool.html#borrowObject-K-) I saw the following.

```
IllegalStateException - after close has been called on this pool
Exception - when makeObject throws an exception
NoSuchElementException - when the pool is exhausted and cannot or will not return another instance
```

**Comment by Manno15** [comment]
@phrocker Do you know the status of this PR?

**Comment by phrocker** [comment]
@Manno15  I see there are changes that I can resolve; however, I haven't encountered these cases in production recently so I can't say if this is globally still an issue that is pervasive. (Number of open sources and aka the compressors not being returned ). 

Is this an issue you have encountered? I will update this, but would like to validate using the tests here on main that this is indeed an issue, still, otherwise I will close it.

**Comment by Manno15** [comment]
No, I haven't encountered this issue in particular. I was just looking through some of the old PR's we have on accumulo to see if any that I can review or complete.

**Review by sirpenev — APPROVED** [review]

**Comment by milleruntime** [comment]
@ivakegg you approved this PR way back in 2017 because I assume it was relevant to a production issue. Are you still seeing issues with this?

**Comment by milleruntime** [comment]
@phrocker I was looking at this PR and took the liberty of merging some conflicts. I made a PR to your branch.

**Comment by milleruntime** [comment]
@phrocker any idea what version of hadoop this was originally developed for? I am trying to figure out if `CodecPool` changed and if this is still a problem.
https://hadoop.apache.org/docs/r2.4.1/api/org/apache/hadoop/io/compress/CodecPool.html

**Comment by phrocker** [comment]
> @phrocker any idea what version of hadoop this was originally developed for? I am trying to figure out if `CodecPool` changed and if this is still a problem. https://hadoop.apache.org/docs/r2.4.1/api/org/apache/hadoop/io/compress/CodecPool.html

Last time I confirmed the issue was in late 2016 ( the first time I forgot about the PR ) --  I want to say it was 2.7.3, but I cannot be 100% certain.

**Comment by milleruntime** [comment]
After some research into recent Hadoop improvements and since production hasn't encountered memory issues with compressors recently, I think this can be closed. Hadoop has made changes in version 2.9.0 to `CodecPool` that essentially does no pooling, one improvement this PR was making. Here is the code from `CodecPool` that checks for the `DoNotPool` annotation:
https://github.com/apache/hadoop/blob/rel/release-2.9.0/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/io/compress/CodecPool.java#L160-L165

**Comment by phrocker** [comment]
> After some research into recent Hadoop improvements and since production hasn't encountered memory issues with compressors recently, I think this can be closed. Hadoop has made changes in version 2.9.0 to `CodecPool` that essentially does no pooling, one improvement this PR was making. Here is the code from `CodecPool` that checks for the `DoNotPool` annotation: https://github.com/apache/hadoop/blob/rel/release-2.9.0/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/io/compress/CodecPool.java#L160-L165

@milleruntime  thanks for doing that leg work! I suspect this issue wouldn't be encountered while source limiting exists on a production systems. I haven't been part of enough to know to be honest, but I did confirm it in 2020 ( my prior comment was meant to be on a production system ) in the last round of comments with an iterator that deep copied a lot of sources.  I believe that annotation is needed on that class, though. A cursory searched shows that it only the built in built in gzip compressor/decompresor does not pool (https://github.com/apache/hadoop/search?q=DoNotPool) . I think the primary motivation was changing how delegation of compressors worked and to make it pluggable less correcting the initial issue but as Keith identified it may be better suited in spi anyway, which means closing this makes sense.

**Pull request closed without merging**
